pkg/api/pluginproxy/pluginproxy.go

```diff
@@ -5,121 +5,184 @@ import (
 	"encoding/json"
 	"io/ioutil"
 	"net/http"
-	"net/http/httputil"
 	"net/url"
 
+	"github.com/grafana/grafana/pkg/infra/tracing"
 	"github.com/grafana/grafana/pkg/models"
 	"github.com/grafana/grafana/pkg/plugins"
 	"github.com/grafana/grafana/pkg/services/pluginsettings"
 	"github.com/grafana/grafana/pkg/services/secrets"
 	"github.com/grafana/grafana/pkg/setting"
 	"github.com/grafana/grafana/pkg/util"
 	"github.com/grafana/grafana/pkg/util/proxyutil"
+	"github.com/grafana/grafana/pkg/web"
+	"go.opentelemetry.io/otel/attribute"
 )
 
-type templateData struct {
-	JsonData       map[string]interface{}
-	SecureJsonData map[string]string
+type PluginProxy struct {
+	ps             *pluginsettings.DTO
+	pluginRoutes   []*plugins.Route
+	ctx            *models.ReqContext
+	proxyPath      string
+	matchedRoute   *plugins.Route
+	cfg            *setting.Cfg
+	secretsService secrets.Service
+	tracer         tracing.Tracer
+	transport      *http.Transport
 }
 
-// NewApiPluginProxy create a plugin proxy
-func NewApiPluginProxy(ctx *models.ReqContext, proxyPath string, route *plugins.Route,
-	appID string, cfg *setting.Cfg, pluginSettingsService pluginsettings.Service,
-	secretsService secrets.Service) *httputil.ReverseProxy {
-	appProxyLogger := logger.New(
-		"userId", ctx.UserId,
-		"orgId", ctx.OrgId,
-		"uname", ctx.Login,
-		"app", appID,
-		"path", ctx.Req.URL.Path,
-		"remote_addr", ctx.RemoteAddr(),
-		"referer", ctx.Req.Referer(),
-	)
+// NewPluginProxy creates a plugin proxy.
+func NewPluginProxy(ps *pluginsettings.DTO, routes []*plugins.Route, ctx *models.ReqContext,
+	proxyPath string, cfg *setting.Cfg, secretsService secrets.Service, tracer tracing.Tracer,
+	transport *http.Transport) (*PluginProxy, error) {
+	return &PluginProxy{
+		ps:             ps,
+		pluginRoutes:   routes,
+		ctx:            ctx,
+		proxyPath:      proxyPath,
+		cfg:            cfg,
+		secretsService: secretsService,
+		tracer:         tracer,
+		transport:      transport,
+	}, nil
+}
 
-	director := func(req *http.Request) {
-		query := pluginsettings.GetByPluginIDArgs{OrgID: ctx.OrgId, PluginID: appID}
-		ps, err := pluginSettingsService.GetPluginSettingByPluginID(ctx.Req.Context(), &query)
-		if err != nil {
-			ctx.JsonApiErr(500, "Failed to fetch plugin settings", err)
-			return
+func (proxy *PluginProxy) HandleRequest() {
+	// found route if there are any
+	for _, route := range proxy.pluginRoutes {
+		// method match
+		if route.Method != "" && route.Method != "*" && route.Method != proxy.ctx.Req.Method {
+			continue
 		}
 
-		secureJsonData, err := secretsService.DecryptJsonData(ctx.Req.Context(), ps.SecureJSONData)
-		if err != nil {
-			ctx.JsonApiErr(500, "Failed to decrypt plugin settings", err)
-			return
+		if !web.MatchTest(route.Path, proxy.proxyPath) {
```

Comment:
Think I have to use Tree.Match to get the matched `params` and pick out the `*` param and set the proxy path to that one.
https://github.com/grafana/grafana/blob/db9c9b5354e0e58fbc733fe11a6f229ab6be6dd8/pkg/web/tree.go#L377